lib/src/local_storage.dart

```diff
@@ -0,0 +1,110 @@
+import 'dart:convert';
+
+import 'package:hive_flutter/hive_flutter.dart';
+
+const _hiveBoxName = 'supabase_authentication';
+const supabasePersistSessionKey = 'SUPABASE_PERSIST_SESSION_KEY';
+
+/// LocalStorage is used to persist the user session in the device.
+///
+/// See also:
+///
+///   * [SupabaseAuth], the instance used to manage authentication
+///   * [EmptyLocalStorage], used to disable session persistance
+///   * [HiveLocalStorage], that implements Hive as storage method
+abstract class LocalStorage {
+  const LocalStorage({
+    required this.initialize,
+    required this.hasAccessToken,
+    required this.accessToken,
+    required this.persistSession,
+    required this.removePersistedSession,
+  });
+
+  /// Initialize the storage to persist session.
+  final Future<void> Function() initialize;
+
+  /// Check if there is a persisted session.
+  final Future<bool> Function() hasAccessToken;
+
+  /// Get the access token from the current persisted session.
+  final Future<String?> Function() accessToken;
+
+  /// Remove the current persisted session.
+  final Future<void> Function() removePersistedSession;
+
+  /// Persist a session in the device.
+  final Future<void> Function(String) persistSession;
+}
+
+/// A [LocalStorage] implementation that does nothing. Use this to
+/// disable persistance.
+class EmptyLocalStorage extends LocalStorage {
+  /// Creates a [LocalStorage] instance that disables persistance
+  const EmptyLocalStorage()
+      : super(
+          initialize: _initialize,
+          hasAccessToken: _hasAccessToken,
+          accessToken: _accessToken,
+          removePersistedSession: _removePersistedSession,
+          persistSession: _persistSession,
+        );
+
+  static Future<void> _initialize() async {}
+  static Future<bool> _hasAccessToken() => Future.value(false);
+  static Future<String?> _accessToken() => Future.value(null);
+  static Future<void> _removePersistedSession() async {}
+  static Future<void> _persistSession(_) async {}
+}
+
+/// A [LocalStorage] implementation that implements Hive as the
+/// storage method.
+class HiveLocalStorage extends LocalStorage {
+  /// Creates a LocalStorage instance that implements the Hive Database
+  const HiveLocalStorage()
+      : super(
+          initialize: _initialize,
+          hasAccessToken: _hasAccessToken,
+          accessToken: _accessToken,
+          removePersistedSession: _removePersistedSession,
+          persistSession: _persistSession,
+        );
+
+  /// The encryption key used by Hive. If null, the box is not encrypted
+  ///
+  /// This value should not be redefined in runtime, otherwise the user may
+  /// not be fetched correctly
+  ///
+  /// See also:
+  ///
+  ///   * <https://docs.hivedb.dev/#/advanced/encrypted_box?id=encrypted-box>
+  static String? encryptionKey;
```

Comment:
How to add a encryption key?

```dart
HiveLocalStorage.encryptionKey = MY_ENCRYPTION_KEY;
```

See <https://docs.hivedb.dev/#/advanced/encrypted_box?id=encrypted-box> for more info on how to store your encryption key